### PR TITLE
Max java aggregation framework

### PR DESCRIPTION
**Pull request by MaBeuLux88**

### CODE REVIEW

**Comment by jyemin** [comment]
Did you mean to have the first two commits in this PR?  They are identical to the ones in https://github.com/mongodb-developer/java-quick-start/pull/4.

**Comment by MaBeuLux88** [comment]
I branched from the first branch to avoid conflicting on merges. Sorry
about that :-). Just check the one with all files if you like. Files will
be the same anyway in the end. Thanks

Le lun. 13 janv. 2020 à 15:41, Jeff Yemin <notifications@github.com> a
écrit :

File: src/main/java/com/mongodb/quickstart/AggregationFramework.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+     * find the 3 most densely populated cities in Texas.
+     * @param zips sample_training.zips collection from the MongoDB Sample Dataset in MongoDB Atlas.
+     */
+    private static void threeMostPopulatedCitiesInTexas(MongoCollection<Document> zips) {
+        Bson match = match(eq("state", "TX"));
+        Bson group = group("$city", sum("totalPop", "$pop"));
+        Bson project = project(fields(excludeId(), include("totalPop"), computed("city", "$_id")));
+        Bson sort = sort(descending("totalPop"));
+
+        ArrayList<Document> results = zips.aggregate(Arrays.asList(match, group, project, sort, limit(3)))
```
Comment (jyemin) [inline src/main/java/com/mongodb/quickstart/AggregationFramework.java]: Be consistent: either inline all the stages or inline none of them.  If majority rules, than pull the `limit(3)` out into a variable.

Comment (jyemin) [inline src/main/java/com/mongodb/quickstart/AggregationFramework.java]: Use `List<Document>` as the variable type.  No need to mention the concrete type twice.

File: src/main/java/com/mongodb/quickstart/AggregationFramework.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+     * find the 3 most popular tags and their post titles
+     * @param posts sample_training.posts collection from the MongoDB Sample Dataset in MongoDB Atlas.
+     */
+    private static void threeMostPopularTags(MongoCollection<Document> posts) {
+        Bson unwind = unwind("$tags");
+        Bson group = group("$tags", sum("count", 1L), push("titles", "$title"));
+        Bson sort = sort(descending("count"));
+        Bson project = project(fields(excludeId(), computed("tag", "$_id"), include("count", "titles")));
+
+        ArrayList<Document> results = posts.aggregate(Arrays.asList(unwind, group, sort, limit(3), project)).into(new ArrayList<>());
```
Comment (jyemin) [inline src/main/java/com/mongodb/quickstart/AggregationFramework.java]: Same as above re: limit stage

File: src/main/java/com/mongodb/quickstart/MappingPOJO.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            Grade newGrade = new Grade().setStudent_id(10003d)
+                                        .setClass_id(10d)
+                                        .setScores(singletonList(new Score().setType("homework").setScore(50d)));
+            grades.insertOne(newGrade);
+
+            // find this grade.
+            Grade grade = grades.find(eq("student_id", 10003d)).first();
+            System.out.println("Grade found:\t" + grade);
+
+            // update this grade: adding an exam grade
```
Comment (jyemin) [inline src/main/java/com/mongodb/quickstart/MappingPOJO.java:53]: Sure we don't just want to do an update and push a new score onto the array?

File: src/main/java/com/mongodb/quickstart/models/Grade.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    public ObjectId getId() {
+        return id;
+    }
+
+    public Grade setId(ObjectId id) {
+        this.id = id;
+        return this;
+    }
+
+    public Double getStudent_id() {
```
Comment (jyemin) [inline src/main/java/com/mongodb/quickstart/models/Grade.java]: From a Java perspective this is a non-standard naming convention.  Consider `getStudentId` and use `@BsonProperty` if you want the field name to be different in the database itself.

**Review by jyemin — CHANGES_REQUESTED** [review]

File: src/main/java/com/mongodb/quickstart/MappingPOJO.java (hunk 3 above)
Comment (MaBeuLux88) [reply src/main/java/com/mongodb/quickstart/MappingPOJO.java:53]: If I do that then I have to manipulate "Document" again. I covered this in the update blog post. Here I'm trying to show how to make it using only POJOs. 
But yes, it's less optimal using POJOs, I totally agree but that's the trade off for using POJOs in this case of situation.

File: src/main/java/com/mongodb/quickstart/models/Grade.java (hunk 4 above)
Comment (MaBeuLux88) [reply src/main/java/com/mongodb/quickstart/models/Grade.java]: Updated + I will add a comment in the blog post about it.

File: src/main/java/com/mongodb/quickstart/AggregationFramework.java (hunk 2 above)
Comment (MaBeuLux88) [reply src/main/java/com/mongodb/quickstart/AggregationFramework.java]: done

File: src/main/java/com/mongodb/quickstart/AggregationFramework.java (hunk 1 above)
Comment (MaBeuLux88) [reply src/main/java/com/mongodb/quickstart/AggregationFramework.java]: Automatic IntelliJ "store in variable feature"... Maybe some day they will get it right and use a List ^^.

Comment (MaBeuLux88) [reply src/main/java/com/mongodb/quickstart/AggregationFramework.java]: done

**Pull request merged**